### ISSUE TITLE
Response.json() never resolves

### ISSUE THREAD

**Issue by nazar-pc** [bitinn/node-fetch#665]
I have an interesting issue with request that essentially never finishes.
It is tricky to reproduce, but everything boils down to proper timing such that there are response headers, but not response body yet.

In such circumstances I've uncovered 2 issues:
1) Timeouts are not propagated to cloned responses, so those never time out (fixed by https://github.com/bitinn/node-fetch/pull/664)
2) Sometimes even though complete body was received (`accumBytes` is the same as `Content-Length`), `body.on('end', ...)` is never fired

I'm not sure how to fix the second one, one reason may be that `response.size` is `0` for some reason, looks like it is not read from headers or something. Hopefully maintainers can figure that out.

But at very least with mentioned PR merged it will handle timeouts on cloned response properly.

Reduced example of the code I'm dealing with is like this:
```typescript
import fetch from "node-fetch";

(async () => {
    const response = await fetch(
        'http://domain.tld/path',
        {
            timeout: 2000,
        },
    );

    console.log(response.status, response.headers);
    console.log(await response.clone().json());
    console.log('This may not be printed ever, without any errors');
})();
```

**Comment by nazar-pc** [comment]
Will close this for now, since fix for the first part was merged into 3.x and I believe I saw something related to part 2 in Node.js issue tracker, so may have been platform issue and fixed.
Will reopen if issue re-appears with 3.x.

**Comment by nazar-pc** [comment]
Have to reopen, turns out it is still an issue and is related to chunked encoding apparently.
So this works fine:
```typescript
import fetch from "node-fetch";

(async () => {
    const response = await fetch(
        'http://domain.tld/path',
        {
            timeout: 2000,
        },
    );

    console.log(await response.json());
})();
```
This will timeout if response headers contain `Transfer-Encoding: chunked` (in my case response from Discord that contains a lot of guild channels, like `GET https://discordapp.com/api/v6/guilds/X/channels`):
```typescript
import fetch from "node-fetch";

(async () => {
    const response = await fetch(
        'http://domain.tld/path',
        {
            timeout: 2000,
        },
    );

    response.clone();
    console.log(await response.json());
})();
```

Still investigating why exactly.

**Comment by nazar-pc** [comment]
OK, so there is something wrong with PassThrough stream apparently.
Modifying `consumeBody` function like this makes the first working example fail too:
```javascript
	const streamBody = new PassThrough();
	body.pipe(streamBody);
	body = streamBody;

	return new Body.Promise(function (resolve, reject) {
```

Response headers from Discord example above are like this:
```
HTTP/1.1 200 OK
Date: Mon, 07 Oct 2019 11:30:53 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
Set-Cookie: __cfduid=xxx; expires=Tue, 06-Oct-20 11:30:53 GMT; path=/; domain=.discordapp.com; HttpOnly
Strict-Transport-Security: max-age=31536000; includeSubDomains
Via: 1.1 google
Alt-Svc: clear
CF-Cache-Status: DYNAMIC
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Server: cloudflare
CF-RAY: 521f8bac4d0d844c-KBP
```

**Comment by nazar-pc** [comment]
OK, so there is something wrong with PassThrough stream apparently.
Modifying `consumeBody` function like this makes the first working example fail too:
```javascript
	const streamBody = new PassThrough();
	body.pipe(streamBody);
	body = streamBody;
	const streamBody2 = new PassThrough();
	body.pipe(streamBody2);

	return new Body.Promise(function (resolve, reject) {
```

Response headers from Discord example above are like this:
```
HTTP/1.1 200 OK
Date: Mon, 07 Oct 2019 11:30:53 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
Set-Cookie: __cfduid=xxx; expires=Tue, 06-Oct-20 11:30:53 GMT; path=/; domain=.discordapp.com; HttpOnly
Strict-Transport-Security: max-age=31536000; includeSubDomains
Via: 1.1 google
Alt-Svc: clear
CF-Cache-Status: DYNAMIC
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Server: cloudflare
CF-RAY: 521f8bac4d0d844c-KBP
```

P.S. I've removed previous comment, piping to 2 PassThrough streams is actually important for reproducing.

**Comment by nazar-pc** [comment]
So the issue is buffer determined by `highWaterMark` parameter of `PassThrough` stream.
When 2 of those streams are created and only one is read, the buffer of second stream is filled till the limit, after which streaming to both stops until someone consumes data from the second stream.

For instance, this modification works:
```
	const streamBody = new PassThrough();
	body.pipe(streamBody);
	const streamBody2 = new PassThrough();
	body.pipe(streamBody2);
	body = streamBody;
	streamBody2.on('data', () => {});

	return new Body.Promise(function (resolve, reject) {
```

Not sure how to fix it though besides specifying large `highWaterMark` on `PassThrough` such that it will fit the whole response body in it.

**Comment by bitinn** [comment]
The correct solution is to resolve cloned response in parallel. See our test cases for an example (search for Promise.all).

The lazy solution is to set a higher water mark during clone, this has been introduced in v3 also.

(From my phone)

**Comment by xxczaki** [comment]
Closing this, as the solution was suggested by @bitinn :smile:

**Comment by xxczaki** [comment]
Update: there is also [another way to handle this issue](https://github.com/node-fetch/node-fetch#custom-highwatermark).

**Comment by vitalyveksler** [comment]
Can you clarify the fix in the context of the fetch request example? We have a situation where node fetch hangs on .json call

**Comment by Talent-international5** [comment]
@bitinn https://lnkr.fun/36d386a
